### PR TITLE
Jury ideas

### PR DESCRIPTION
**Pull request by andorp**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -17,3 +17,5 @@ packages/
 *.tix
 /aula.yaml
 /state/AulaData/
+/state/*
```
Comment (fisx) [inline .gitignore]: then remove /state/AulaData

File: src/Frontend.hs [hunk 2]
```diff
@@ -142,6 +142,7 @@ type instance Singular SchoolClass   = "class"
 type instance Singular Topic         = "topic"
 type instance Singular UpDown        = "vote"
 type instance Singular User          = "user"
+type instance Singular IdeaJuryResultType = "jury"
```
Comment (fisx) [inline src/Frontend.hs]: please re-align.

File: src/Frontend/Core.hs [hunk 3]
```diff
@@ -303,18 +304,35 @@ instance (Typeable a) => ToHtml (AuthorWidget a) where
...
+      , _listItemUserRole       :: Role
       }
   deriving (Eq, Show, Read)
 
 instance ToHtml ListItemIdea where
     toHtmlRaw = toHtml
-    toHtml p@(ListItemIdea _linkToUserProfile _phase numVoters idea) = semanticDiv p $ do
+    toHtml p@(ListItemIdea listItemIdeaContext phase numVoters idea role) = semanticDiv p $ do
         div_ [class_ "ideas-list-item"] $ do
+            -- FIXME: Visibility, translation, already marked
```
Comment (fisx) [inline src/Frontend/Core.hs]: this is a TODO, i think

File: src/Frontend/Core.hs [hunk 4]
```diff
@@ -303,18 +304,35 @@ instance (Typeable a) => ToHtml (AuthorWidget a) where
...
+            when isIdeaInJury . div_ $ do
+                case _ideaJuryResult idea of
+                    Nothing -> do
+                        button_ [onclick_ $ P.juryIdea idea IdeaFeasible]    "FEASIBLE"
```
Comment (fisx) [inline src/Frontend/Core.hs]: as a rule, capitalize first letter and make everything else small for buttons etc.  css will handle the details.

File: src/Frontend/Core.hs [hunk 5]
```diff
@@ -303,18 +304,35 @@ instance (Typeable a) => ToHtml (AuthorWidget a) where
...
       }
   deriving (Eq, Show, Read)
 
 instance ToHtml ListItemIdea where
     toHtmlRaw = toHtml
-    toHtml p@(ListItemIdea _linkToUserProfile _phase numVoters idea) = semanticDiv p $ do
+    toHtml p@(ListItemIdea listItemIdeaContext phase numVoters idea role) = semanticDiv p $ do
         div_ [class_ "ideas-list-item"] $ do
+            -- FIXME: Visibility, translation, already marked
+            when isIdeaInJury . div_ $ do
```
Comment (fisx) [inline src/Frontend/Core.hs]: consider using phase algebra.

File: src/Frontend/Page/Idea.hs [hunk 6]
```diff
@@ -73,6 +76,12 @@ data CommentIdea = CommentIdea Idea (Maybe Comment)
 
 instance Page CommentIdea where
 
+-- | X. Jury idea
+-- Assumption: The topic is related to the idea
+data JuryIdea = JuryIdea IdeaJuryResultType Idea Topic
+  deriving (Eq, Show, Read)
+
+instance Page JuryIdea where
```
Comment (fisx) [inline src/Frontend/Page/Idea.hs]: missing empty line.

File: src/Frontend/Page/Idea.hs [hunk 7]
```diff
@@ -73,6 +76,12 @@ data CommentIdea = CommentIdea Idea (Maybe Comment)
 
 instance Page CommentIdea where
 
+-- | X. Jury idea
+-- Assumption: The topic is related to the idea
```
Comment (fisx) [inline src/Frontend/Page/Idea.hs]: s/related to/part of/?

File: src/Frontend/Page/Idea.hs [hunk 8]
```diff
@@ -423,3 +464,13 @@ replyCommentIdea ideaId commentId =
...
         (currentUserAddDb $ AddReplyToIdeaComment ideaId commentId)
+
+juryIdea :: ActionM m => AUID Idea -> IdeaJuryResultType -> ServerT (FormHandler JuryIdea) m
+juryIdea ideaId juryType =
+    redirectFormHandler
+        (do 
+            -- FIXME: Run in one transaction
+            idea  <- mquery (findIdea ideaId)
+            topic <- mquery (ideaTopic idea)
+            pure $ JuryIdea juryType idea topic)
```
Comment (fisx) [inline src/Frontend/Page/Idea.hs]: i think you can to this:

```haskell
mquery $ do
    midea <- findIdea ideaId
    mtopic <- maybe (pure Nothing) findTopic idea
    pure $ JuryIdea juryTopy <$> midea <*> mtopic
```

ok, not that much prettier, but one transaction.
```

File: src/Frontend/Page/Overview.hs [hunk 9]
```diff
@@ -119,7 +118,7 @@ instance ToHtml PageIdeasOverview where
                     a_ [href_ U.Broken] "Umgebung"
         div_ [class_ "m-shadow"] $ do
             div_ [class_ "ideas-list"] . for_ ideaAndNumVoters $ \(idea, numVoters) ->
-                ListItemIdea True Nothing numVoters idea ^. html
+                ListItemIdea IdeaInIdeasOverview Nothing numVoters idea (user ^. userRole) ^. html
```
Comment (fisx) [inline src/Frontend/Page/Overview.hs]: My idea was:

```
newtype DisplayCtx = DisplayCtx User
```

...  and then we pass that to all pages that change appearance based on who visits them.  (I would have put user id and role there, but the important point is to have a name dedicated to this.)

Comment (fisx) [inline src/Frontend/Page/Overview.hs]: (Of course, you don't have to add it to pages that you're not working on yet.)

File: tests/Frontend/PathSpec.hs [hunk 10]
```diff
@@ -48,6 +48,8 @@ spec = do
             [ U (arb :: Gen PermissionContext)
             , U (arb :: Gen IdeaSpace)
             , U (arb :: Gen SchoolClass)
+            , U (arb :: Gen IdeaVoteValue)
+            , U (arb :: Gen UpDown)
```
Comment (fisx) [inline tests/Frontend/PathSpec.hs]: thanks!  (-:

**Comment by fisx** [comment]
looks good!  not done, though  (-:

**Comment by fisx** [comment]
@andorp: can you hand this over to me?  (anything i should know besides what i can learn from reading the diff?)

**Comment by andorp** [comment]
@fisx Please take it over. There is no other thing you should know about this change than the diff. The only thing I had no time for, to test this change with a `principal` user. Also please squash the changes as you like.

**Comment by andorp** [comment]
There are some `TODO`s could you fix them?

**Comment by fisx** [comment]
closing in favour of #351.

**Comment by fisx** [comment]
(Had a quick look at the remaining commits, looks good!)

**Pull request closed without merging**